README.md
cocotte-is
======

node.js変数チェック用ライブラリ

jsに静的型指定機能を追加する為の簡易ライブラリです

underscoreの補完を目的として作成されているため、
underscoreに同等の機能が存在する場合はそちらを優先して使用してください

ただしcocotte-isはunderscoreには依存していません
単体で動作します

#使用方法

```javascript
var cis = require('cocotte-is');

var fn = function (p1) {
    if (!cis(String, p1)) {
        throw new TypeError('引数1が文字列ではありません');
    }
};
```

#関数

##cis

型判定を行います  

`cis(型, 値)`と使用し、真偽値を返します  
型は基本的に値の`constructor`と一致した時に`true`です

ただし、いくつかの値に関しては型はconstructor以外になります

```javascript
var x = cis(String, 'abc');
var y = cis(Infinity, 1/0);
```

`instanceof`に似ていますが、プロトタイプチェーンをたどりません  
例えば`new String('abc')`を`String`であるかつ`Object`ではないと判別するには  
`instanceof`の代わりに`cis`を使用する必要があります  
また`null`, `NaN`などを判別する場合も使用できます  

値の`constructor`ではないものは以下の通りです

| 値                      | true              | false          |
| ----------------------- | ------------------| -------------- |
| undefined               | cis(undefined, x) | cis(null  , x) |
| null                    | cis(null     , x) | cis(Object, x) |
| NaN                     | cis(NaN      , x) | cis(Number, x) |
| Infinity                | cis(Infinity , x) | cis(Number, x) |
| Number.POSITIVE_INFINITY| cis(Infinity , x) | cis(Number, x) |
| Number.NEGATIVE_INFINITY| cis(-Infinity, x) | cis(Number, x) |
| new Data('abcdefg')     | cis(NaN      , x) | cis(Date  , x) |

（注意）`undefined`は安全に検証する為の`cis.undefined(x)`を使用してください

##cis.be

型判定を行います  
`cis`と似ていますが、入力値が`null`または`undefined`の場合は`true`を返します

```javascript
var x = cis.be(String, 'abc');
var y = cis.be(String, null);
```

どちらも`true`を返します

##cis.alt

型判定を行い、一致した場合は値をそのまま返し、不一致の場合は代替の値を返します  
第三引数を省略した場合は、代替の値は`null`です

```javascript
var x = cis.alt(String, 'abc', 'def'); // x -> 'abc'
var y = cis.alt(Number, '123', 456);   // y -> 456
var z = cis.alt(Boolean, 't');         // z -> null
```

##cis.getType
型を返します

```javascript
var type = cis.getType('abc');
```

##cis.enableId 
識別子として利用可能かどうかを確認します

```javascript
var x = cis.enableId('xyz');
var y = cis.enableId('xyz', ['abc', 'xyz']);
```

識別子は予約語ではなく、32文字以内の英数字である必要があります  
また、第2引数に追加の予約語を設定する事が出来ます

##cis.reservedWords
予約語一覧を取得します

```javascript
var ls = cis.reservedWords();
```

##cis.undefined
対象が未設定の値(`undefined`)の場合に`true`を返します  
`undefined`を安全に検証する事ができます

```javascript
var x = cis.undefined(123456);
```

##cis.integer
対象が整数で`-2147483648〜2147483647`の範囲の場合に`true`を返します  
(64bit環境でも範囲は同じです)  
文字列は数値の表現でも`false`を返します  
`1.00`等小数点以下が0の値も整数と見なします

```javascript
var x = cis.integer(123456);
```

##cis.unset
対象が`null`、`undefined`の場合に`true`を返します  
空文字や数値の0と厳密に区別を行いたい場合に使用してください  

```javascript
var x = cis.unset(0);
```

##cis.arg
対象が`Arguments`オブジェクトの場合はtrueを返します  
実際には`Arguments`は定義がないですが、`cis(Arguments, x)`のように動作します  
`_.cisArguments`と同じです  

```javascript
var x = cis.arg(arguments);
```

##cis.error
エラーオブジェクトかどうかを確認します  
この関数はプロトタイプチェーンをさかのぼります  
`cis`との違いは`cis(Error, new TypeError('x'))`は`false`になる事に対し
下記は`true`になります

```javascript
var x = cis.error(new TypeError('文字列ではありません'));
```

##cis.between
値が範囲内かどうかを確認します  
引数はすべて数字もしくは`null`である必要があります

下記は、`x`が`1`から`10`の間に収まっているかを判別します  
開始もしくは終了の値が`null`の場合は、以上・以下の判別です

```javascript
var x = 5;
var y = cis.between(1, x, 10);
```

##cis.allString
すべての値が文字列であるか判別します

```javascript
var x = cis.allString(['a', 'b', 'c']);
```

##cis.allNumber
すべての値が有効な数字であるか判別します  
`NaN`,`Infinity`, `Number.NEGATIVE_INFINITY`は無効な数字と見なします

```javascript
var x = cis.allString([1, 2, 3]);
```

##cis.all
すべての値が指定した型の要素の配列かどうかを判別します  

```javascript
var x = cis.all(Date, [new Date(), new Date(), '2014-01-12']);
```

##cis.unique
すべての値が一意であるか判別します  
各要素の判定は`===`で行われます

```javascript
var x = cis.unique([1, 2, 3]);
```

##cis.matches
すべての型が一致するかの判定を行います

  + @method matches
  + @static
  + @param  {Array}   types
  + @param  {Array|Arguments}   target
  + @param  {Object}  must  省略可能
    + 省略出来ない値のインデックス設定していない場合はすべて省略不可とします
    + 省略可の値はcis.unsetがtrueでも判定をパスします
  + @return {Boolean} match

```javascript
var fn = function (p1, p2, p3) {
    if( ! cis.matches([String, Number, Array], arguments) {
         throw new TypeError('引数の型が一致しません');
    }
};
```







History.md
0.2.12 / 2015-03-24
===================

* is.interface削除

0.2.11 / 2014-05-19
===================

 * cis.integer 追加

0.2.10 / 2014-05-17
===================

  * cis.be 追加

0.2.9 / 2014-04-30
===================

  * constructorの取得をプロトタイプに変更
  * テストコード充実
  * readme.mdの修正

0.2.8 / 2014-04-28
===================

  * isが将来予約語になる可能性を考慮してcisに変更
  * 予約語 追加
  * cis.alt 追加

0.2.7 / 2014-04-26
===================

  * is.all 追加

0.2.6 / 2014-04-23
===================

  * use strictを文頭にまとめる
  * コーディング規約に順序。タブをスペースに
  * is NaNとの判定を簡略化
  * is.unique NaNに対応
  * is.between Infinityに対応
  * is.interface 例外を発行する代わりにコールバック関数を引数に設定するように変更
  * README.mdを大幅加筆

0.2.5 / 2014-04-14
===================

  * is 速度重視のリファクタリング
  * is.getType 速度重視のリファクタリング

0.2.4 / 2014-04-03
===================

  * is.enableId 引数callbackを廃止

0.2.3 / 2014-03-30
===================

  * is.undefined undefinedを安全に比較するメソッド追加
  * その他のいくつかの関数: undefinedとの比較をvoid 0との比較に変更

0.2.2 / 2013-10-14
===================

  * is.enableId reservedWords省略時のcallback指定に対応

0.2.0 / 2013-10-08
===================

  * is.allString 追加
  * is.allNumber 追加
  * is.between 追加
  * is.matches 変更。オプションを省略時引数のみ変更
  * is.intefaceCheck追加

0.1.0 / 2013-10-03
===================

  * is.enableId 追加
  * is.reservedWords 追加
  * is.getType 追加
  * is.unset 追加
  * is.arg 追加
  * is.error 追加
  * is.unique 追加
  * is.matches 追加

0.0.1 / 2013-10-01
===================

  * is 追加




test/be.js
'use strict';
var assert = require('assert');
var cis = require('..');

assert( cis.be(String, void 0));
assert( cis.be(String, null));
assert( cis.be(String, 'foo'));
assert(!cis.be(String, 123));

test/is.js
'use strict';
var assert = require('assert');
var cis = require('..');

var Fn = function Fn (){};

// null/undefined
assert(cis(undefined, void 0));
assert(cis(null, null));
assert(!cis(null, 0));
assert(!cis(null, ''));
assert(!cis(null, false));
assert(!cis(null, NaN));

// string
assert(!cis(String, void 0));
assert(!cis(String, null));
assert(cis(String, 'abc'));
assert(cis(String, ''));
assert(cis(String, new String()));
assert(!cis(String, 1));
assert(!cis(String, new Number()));
assert(!cis(String, true));
assert(!cis(String, new Boolean()));

// number
assert(!cis(Number, void 0));
assert(!cis(Number, null));
assert(cis(Number, 1));
assert(cis(Number, -1));
assert(cis(Number, 0));
assert(cis(Number, new Number()));
assert(!cis(Number, '1'));
assert(!cis(Number, NaN));
assert(!cis(Number, Infinity));
assert(!cis(Number, Number.POSITIVE_INFINITY));
assert(!cis(Number, Number.NEGATIVE_INFINITY));

// NaN/Infinity/-Inifinity
assert(cis(NaN, NaN));
assert(cis(NaN, new Date('abcde')));
assert(cis(Infinity, Infinity));
assert(cis(Infinity , Number.POSITIVE_INFINITY));
assert(cis(Number.NEGATIVE_INFINITY, -Infinity));
assert(cis(-Infinity, Number.NEGATIVE_INFINITY));

// Date
assert(cis(Date, new Date()));
assert(!cis(Date, new Date('unknown time')));

// other
assert(cis(Object, {}));
assert(!cis(Object, null));
assert(cis(Function, Fn));
assert(cis(Fn, new Fn()));
assert(!cis(Function, new Fn()));


test/all.js
'use strict';
var assert = require('assert');
var all = require('..').all;

assert(all(Date, []));
assert(!all(Date, [undefined]));
assert(all(Date, [new Date(), new Date(), new Date()]));
assert(all(Date, new Array()));
assert(all(Date, new Array(new Date(), new Date(), new Date())));
assert(all(Number, new Array(10)));
assert(!all(Date, null));
assert(!all(Date, [new Date(), new Date(), '2000-1-1']));
assert(!all(Date, ['a', 'b', 'c']));
assert(!all(Date, new Date()));




bench/old-is.js
'use strict';

var cis = require('..');

var oldIs = function oldIs (type, target) {

  if(target === void 0 || target === null) {
    return type === target;

  } else {
    var ctor = target.constructor;

    if (ctor === Number && !isFinite(target)) {
      return type === target || Number.isNaN(type) && Number.isNaN(target);

    } else if (ctor === Date && isNaN(target.getTime())) {
      return Number.isNaN(type);

    } else {
      return type === ctor;

    }
  }
};


var t0 = (new Date()).getTime();

for(var i=0; i < 100000000; i++) {
  cis(Boolean, true);
}

var t1 = (new Date()).getTime();
console.log('current version: ' + (t1 - t0));

for(var i=0; i < 100000000; i++) {
  oldIs(Boolean, true);
}

var t2 = (new Date()).getTime();
console.log('old version: ' + (t2 - t1));


var assert = require('assert');

assert(t1 - t0 < t2 - t1);

console.log('test ok');

cis.js
/*
 * Copyright(c) 2013 Yuki Kurata <yuki.kurata@gmail.com>
 * MIT Licensed
 */
'use strict';

/** 
 * 型判定機能を追加する為の簡易ライブラリです
 *
 * underscoreの補完を目的として作成されているため
 * underscoreに同等の機能が存在する場合はそちらを優先して使用してください
 * cocotte-isはunderscoreには依存していません
 * 単体で動作します
 */

/*
 * alias
 */
// 識別に使用するID
var regId = /^[a-z]([_0-9a-z]{0,30}[0-9a-z])?$/i;

/**
 * 型判定を行う
 * instanceofに似ていますが、プロトタイプチェーンをたどりません
 * 
 * 例えばnew String('abc')をStringであるかつObjectではないと判別するには
 * instanceofの代わりにisを使用する必要があります
 * 
 * またnullやNaNを判別する場合も使用できます
 *
 * instanceofとは異なる判定方法を行うものを次にまとめます
 * 
 *    undefined               : cis(undefined, x) -> true 
 *          (確実に判定する為にcis.undefinedを使用してください)
 *    null                    : cis(null     , x) -> true
 *    NaN                     : cis(NaN      , x) -> true
 *    Infinity                : cis(Infinity , x) -> true
 *    Number.POSITIVE_INFINITY: cis(Infinity , x) -> true
 *    Number.NEGATIVE_INFINITY: cis(-Infinity, x) -> true
 *    new Data('abcdefg')     : cis(NaN      , x) -> true
 *
 * それぞれ、Number/Dateと型判定時にfalseになります
 * 
 * @for cis
 * @method cis
 * @static
 * @param  {Mixed}   type
 * @param  {Mixed}   target
 * @return {Boolean} isType
 */
var cis = function cis(type, target) {
  // undefined, null
  if(target === void 0 || target === null) {
    return type === target;
  }

  // 文字列
  if (typeof target === 'string') {
    return type === String;
  }

  // 数字
  if (typeof target === 'number') {
    if (isFinite(target)) {
      return type === Number;

    } else if (target !== target) { // NaN
      return  type !== type;

    } else {
      // Infinity, Number.POSITIVE_INFINITY, Number.NEGATIVE_INFINITY
      return type === target;
    }
  }

  // 真偽値
  if (typeof target === 'boolean') {
    return type === Boolean;
  }

  // その他
  var ctor = Object.getPrototypeOf(target).constructor;

  // 不正な日付
  if (ctor === Date && isNaN(target.getTime())) {
    return Number.isNaN(type);
  }

  // コンストラクタとの一致
  return type === ctor;
};

/**
 * 型判定を行う
 * cisと似ていますが、値がnullまたはundefinedの場合もtrueになります
 * @method be
 * @param  {Mixed} type
 * @param  {Mixed} target
 * @return {Boolean} isType
 */
cis.be = function be (type, target) {
  if (target === void 0 || target === null) {
    return true;
  }
  return cis(type, target);
};

/**
 * valueの型がtypeと一致する場合はvalueを返し
 * しない場合はalternativeを返す
 * @method alt
 * @param  {Function} type
 * @param  {Mixed} target
 * @param  {Mixed} alternative
 * @return {Mixed}
 */
cis.alt = function alt (type, target, alternative) {

  if (arguments.length < 3) {
    alternative = null;
  }

  // undefined/null
  if (target === void 0 || target === null) {
    return alternative;
  }

  // 文字列
  if (type === String) {
    if (typeof target === 'string') {
      return target;
    }
    if (target instanceof String) {
      return target.valueOf();
    }
    return alternative;
  }

  // 数字
  if (type === Number) {
    if (typeof target === 'number') {
      return isFinite(target) ? target : alternative;
    }
    if (target instanceof Number) {
      return target.valueOf();
    }
    return alternative;
  }

  // 真偽値
  if (type === Boolean) {
    if (typeof target === 'boolean') {
      return target;
    }
    if (target instanceof Boolean) {
      return target.valueOf();
    }
    return alternative;
  }

  // 日付
  if (type === Date) {
    if (target instanceof Date && !isNaN(target.getTime())) {
      return target;
    }
    return alternative;
  }

  // その他
  return target instanceof type ? target : alternative;
};

/**
 * 型を返します
 *
 * is関数とロジックは似ています
 * ただ、一致を判定するのはなく型をそのまま返します
 *
 * @method getType
 * @static
 * @param  {Mixed}  target
 * @return {Object} type
 */
cis.getType = function getType (target) {

  // undefined, null
  if(target === void 0 || target === null) {
    return target;
  }

  // 文字列
  if (typeof target === 'string') {
    return String;
  }

  // 数字
  if (typeof target === 'number') {
    if (isFinite(target)) {
      return Number;

    } else if (Number.isNaN(target)) {
      return NaN;

    } else {
      // Infinity, Number.POSITIVE_INFINITY, Number.NEGATIVE_INFINITY
      return target;
    }
  }

  // 真偽値
  if (typeof target === 'boolean') {
    return Boolean;
  }

  // その他
  var ctor = Object.getPrototypeOf(target).constructor;

  // 日付
  if (ctor === Date) {
    return isNaN(target.getTime()) ? NaN : Date;
  }

  return ctor;
};

/*
 * 識別子として禁止用語
 * @property {Array} ngWords
 */
var ngWords = [
  // 構文予約語,
  'break', 'delete', 'if', 'this', 'while',
  'case', 'do', 'in', 'throw', 'with',
  'catch', 'else', 'instanceof', 'try', 'continue',
  'finally', 'new', 'typeof', 'debugger', 'for',
  'return', 'var', 'default', 'function', 'switch',
  'void', 'class', 'const', 'enum', 'export',
  'extends', 'import', 'super', 'implements', 'interface',
  'let', 'package', 'yield',
  // ECMAScript,
  'null', 'true', 'false', 'get', 'set', 'arguments',
  // グローバルオブジェクト
  'Array', 'ArrayBuffer', 'Boolean', 'Collator', 'DataView',
  'Date', 'DateTimeFormat', 'decodeURI', 'decodeURIComponent', 'encodeURI',
  'encodeURIComponent', 'Error', 'escape', 'eval', 'EvalError', 'Float32Array',
  'Float64Array', 'Function', 'Infinity', 'Intl', 'Int16Array',
  'Int32Array', 'Int8Array', 'isFinite', 'isNaN', 'Iterator',
  'JSON', 'Math', 'NaN', 'Number', 'NumberFormat',
  'Object', 'parseFloat', 'parseInt', 'RangeError', 'ReferenceError',
  'RegExp','RegExpURIError',  'StopIteration', 'String', 'SyntaxError', 'TypeError',
  'Uint16Array', 'Uint32Array', 'Uint8Array', 'Uint8ClampedArray', 'undefined',
  'unescape', 'uneval', 'URIError',
  // safety
  'private', 'protected', 'public', 'static',
  'abstract', 'as', 'boolean', 'byte', 'char', 'double', 'final',
  'float', 'goto', 'int', 'is', 'long', 'namespace', 'native',
  'short', 'synchronized', 'throws', 'transient', 'use', 'volatile',
  '',
  // node.js,
  'exports', 'require', 'module', '__filename', '__dirname',
  'DTRACE_NET_SERVER_CONNECTION', 'DTRACE_NET_STREAM_END',
  'DTRACE_NET_SOCKET_READ', 'DTRACE_NET_SOCKET_WRITE',
  'DTRACE_HTTP_SERVER_REQUEST', 'DTRACE_HTTP_SERVER_RESPONSE',
  'DTRACE_HTTP_CLIENT_REQUEST', 'DTRACE_HTTP_CLIENT_RESPONSE',
  'global', 'process', 'GLOBAL', 'root', 'Buffer',
  'setTimeout', 'setInterval', 'clearTimeout', 'clearInterval', 'setImmediate',
  'clearImmediate', 'console'
  // user-agent // TODO
].sort();

/**
 * 識別子として利用可能かどうかを確認します
 * @method enableId
 * @param  {String}   target
 * @param  {Array}    optional (省略可能)
 * @return {Boolean}  success
 */
cis.enableId = function enableId (target, optional) {

  if (!cis(String, target)) {
    // 文字列ではない
    return false;

  } else if (~ngWords.indexOf(target)) {
    // システム予約語
    return false;

  } else if (cis.allString(optional) && ~optional.indexOf(target)) {
    // オプションの予約語
    return false;

  } else if (!regId.test(target)) {
    // 32文字以内の英数字ではない
    return false;

  } else {

    return true;
  }
};

/**
 * 予約語一覧を取得する
 * @method reservedWords
 * @param  {Array} optional (省略可能)
 * @return {Array}
 */
cis.reservedWords = function reservedWords (optional) {
  if (cis.allString(optional)) {
    var rtn = [];
    rtn = rtn.concat(ngWords);
    rtn = rtn.concat(optional);
    rtn.sort();
    return rtn;
  } else {
    return ngWords;
  }
};

/**
 * 対象が確実にundefinedであるかどうかを確認する
 * @method undefined
 * @param  {Mixed}  target
 * @return {Boolean}
 */
cis.undefined = function isUndefined (target) {
  return target === void 0;
};

/** 
 * 整数かどうかを確認する
 * @method integer
 * @param  {Mixed}  target
 * @return {Boolean}
 */
if (~~-2147483649 === -2147483649) {
  // 64bit〜
  cis.integer = function integer (target) {
    return typeof target === 'number' && isFinite(target) && target === ~~target &&
      -2147483648 <= target && target <= 2147483647;
  };
} else {
  // 32bit
  cis.integer = function integer (target) {
    return typeof target === 'number' && isFinite(target) && target === ~~target;
  };
}

/**
 * 対象が空文字、null、undefinedの場合にtrueを返します
 * 数値の0やfalseと厳密に区別を行いたい場合に使用してください
 * @method unset
 * @param  {Mixed}   target
 * @return {Boolean}
 */
cis.unset = function isUnset (target) {
  return target === null || target === void 0 || target === '';
};

/** 
 * 対象がArgumentsオブジェクトの場合はtrueを返します
 * 実際にはArgumentsは定義がないですが、cis(Arguments, x)のように動作します。
 * _.isArgumentsと同じ処理です
 * @method arg
 * @param  {Mixed}   target
 * @return {Boolean} isArgumentsObject
 */
cis.arg = function isArg (target) {
  return cis(Object, target) && target.hasOwnProperty('callee');
};

/**
 * エラーオブジェクトかどうかを確認します
 * この関数はプロトタイプチェーンをさかのぼります
 * @method error
 * @param  {target} target
 * @return {Boolaen}
 */
cis.error = function isError (target) {
  return target instanceof Error;
};

/**
 * 値が範囲内かどうかを確認します
 * @method between
 * @param  {Number} min    nullで判定をパス
 * @param  {Number} target 
 * @param  {Number} max    nullで判定をパス
 * @return {Boolean}
 */
cis.between = function isBetween (min, target, max) {

  min = min === null ? -Infinity : min;
  max = max === null ? Infinity : max;

  return typeof min === 'number' &&
         typeof target === 'number' &&
         typeof max === 'number' &&
         min <= target && target <= max;
};

/**
 * すべての値が文字列であるか判別します
 * @method allString
 * @param  {Array}  arr
 * @return {Boolean}
 */
cis.allString = function isAllString (target) {

  if (cis.arg(target)) {
    target = Array.prototype.slice.call(target);

  } else if (!Array.isArray(target)) {
    return false;

  }

  return target.every(function (v) {
    return cis(String, v);
  });
};

/**
 * すべての値が有効な数字であるか判別します
 * @method allNumber
 * @param  {Array}  target
 * @return {Boolean}
 */
cis.allNumber = function isAllNumber (target) {

  if (cis.arg(target)) {
    target = Array.prototype.slice.call(target);

  } else if (!Array.isArray(target)) {
    return false;

  }

  return target.every(function (v) {
    return cis(Number, v);
  });
};

/**
 * すべての値が指定した型の要素の配列かどうかを判別します
 * @method all
 * @param  {Mixed} type
 * @param  {Array} target
 * @return {Boolean}
 */
cis.all = function isAll (type, target) {

  if (cis.arg(target)) {
    target = Array.prototype.slice.call(target);

  } else if (!Array.isArray(target)) {
    return false;

  }

  return target.every(function (v) {
    return cis(type, v);
  });
};

/**
 * すべての値が一意であるか判別します
 * @method unique
 * @param  {Array} target
 * @return {Boolean}
 */
cis.unique = function isUnique (target) {

  if (cis.arg(target)) {
    target = Array.prototype.slice.call(target);

  } else if (!Array.isArray(target)) {
    return false;

  }

  // NaN対策
  var nanCnt = target.reduce(function(x, v){return v !== v ? x + 1: x;}, 0);
  if (nanCnt >= 2) {
    return false;
  }

  return target.every(function (v, i, self) {
    return self.indexOf(v) === self.lastIndexOf(v);
  });

};

/**
 * すべての型が一致するかの判定を行います
 * 
 * @method matches
 * @static
 * @param  {Array}           types
 * @param  {Array|Arguments} target
 * @param  {Array}           must
 *           省略出来ないインデックス。設定していない場合はすべて省略不可とします
 *           省略可の値はcis.unsetがtrueでも判定をパスします
 * @return {Boolean} match
 */
cis.matches = function isMatches (types, target, must) {

  // 対象が配列もしくはArgumentsではない場合はfalseを返す
  if (!Array.isArray(types) && !cis.arg(target)) {
    return false;
  }

  // 判別
  if(Array.isArray(must)) {
    // 省略可能な引数が存在する場合
    return types.every(function (tp, i) {
      return (!~must.indexOf(i) && cis.unset(target[i])) || cis(tp, target[i]);
    });

  } else {
    // すべての引数が必須
    return types.every(function (tp, i) { return cis(tp, target[i]); });

  }
};

module.exports = exports = cis;
